src/refactor.jl

```diff
@@ -0,0 +1,230 @@
+handle("renamerefactor") do data
+  @destruct [
+    old,
+    full,
+    new,
+    path,
+    # local context
+    column || 1,
+    row || 1,
+    startRow || 0,
+    context || "",
+    # module context
+    mod || "Main",
+  ] = data
+  renamerefactor(old, full, new, path, column, row, startRow, context, mod)
+end
+
+function renamerefactor(
+  old, full, new, path,
+  column = 1, row = 1, startrow = 0, context = "",
+  mod = "Main",
+)
+  # catch keyword renaming
+  iskeyword(old) && return Dict(:warning => "Keywords can't be renamed: `$old`")
+
+  mod = getmodule(mod)
+  head = first(split(full, '.'))
+  headval = getfield′(mod, head)
+
+  # catch field renaming
+  head ≠ old && !isa(headval, Module) && return Dict(
+    :warning => "Rename refactoring on a field isn't available: `$obj.$old`"
+  )
+
+  expr = CSTParser.parse(context)
+  items = toplevelitems(expr, context)
+  ind = findfirst(item -> item isa ToplevelBinding, items)
+  bind = ind === nothing ? nothing : items[ind].bind
+
+  # local rename refactor if `old` isn't a toplevel binding
+  if islocalrefactor(bind, old)
+    try
+      refactored = localrefactor(old, new, path, column, row, startrow, context, expr)
+      return isempty(refactored) ?
+        # NOTE: global refactoring not on definition, e.g.: on a call site, will be caught here
+        Dict(:info => contextdescription(old, mod, context)) :
+        Dict(
+          :text    => refactored,
+          :success => "_Local_ rename refactoring `$old` ⟹ `$new` succeeded"
+        )
+    catch err
+      return Dict(:error => errdescription(old, new, err))
+    end
+  end
+
+  # global rename refactor if the local rename refactor didn't happen
+  try
+    kind, desc = globalrefactor(old, new, mod, expr)
+
+    # make description
+    if kind === :success
+      val = getfield′(mod, full)
+      moddesc = if (headval isa Module && headval ≠ mod) ||
+                   (applicable(parentmodule, val) && (headval = parentmodule(val)) ≠ mod)
+        moduledescription(old, headval)
+      else
+        ""
+      end
+
+      desc = join(("_Global_ rename refactoring `$mod.$old` ⟹ `$mod.$new` succeeded.", moddesc, desc), "\n\n")
+    end
+
+    return Dict(kind => desc)
+  catch err
+    return Dict(:error => errdescription(old, new, err))
+  end
+end
```

Comment:
I agree the logic here would look messy, so please feel free to ask me if you have anything unsure -- I probably miss some edge cases.